backend/src/realtime/realtime-note/realtime-note.service.ts

```diff
@@ -109,13 +112,46 @@ export class RealtimeNoteService implements BeforeApplicationShutdown {
   }
 
   @OnEvent(NoteEvent.PERMISSION_CHANGE)
-  public handleNotePermissionChanged(noteId: Note['id']): void {
-    const realtimeNote = this.realtimeNoteStore.find(noteId);
-    if (realtimeNote) {
-      realtimeNote.announcePermissionChange();
+  public async handleNotePermissionChanged(note: Note): Promise<void> {
+    const realtimeNote = this.realtimeNoteStore.find(note.id);
+    if (!realtimeNote) return;
+
+    realtimeNote.announcePermissionChange();
+    const allConnections = realtimeNote.getConnections();
+    await this.closeNoneReadNoteConnections(allConnections, note);
+    await this.updateAcceptEditForAllConnection(allConnections, note);
+  }
+
+  private async updateAcceptEditForAllConnection(
+    connections: RealtimeConnection[],
+    note: Note,
+  ): Promise<void> {
+    for (const connection of connections) {
+      connection.acceptEdits = await this.permissionService.mayWrite(
+        connection.getUser(),
+        note,
+      );
     }
   }
 
+  private async closeNoneReadNoteConnections(
+    connections: RealtimeConnection[],
+    note: Note,
+  ): Promise<void> {
+    for (const connection of connections) {
+      if (!(await this.connectionCanRead(connection, note))) {
+        connection.getTransporter().disconnect();
+      }
+    }
+  }
```

Comment:
You're using the same array of connections for both methods. Therefore if a connection can't read anymore, then you first close it in the first for loop and then check if the user is still allowed to write in the second one. That's a bit redundant. 

But before you start to delete connections from the array.. just handle both cases in one for loop.

Reply:
done